staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/tlsconfig.go

```diff
@@ -94,7 +95,29 @@ func (c *DynamicServingCertificateController) GetConfigForClient(clientHello *tl
 		return nil, errors.New("dynamiccertificates: unexpected config type")
 	}
 
-	return tlsConfig.Clone(), nil
+	tlsConfigCopy := tlsConfig.Clone()
+
+	// if the client set SNI information, just use our "normal" SNI flow
+	if len(clientHello.ServerName) > 0 {
+		return tlsConfigCopy, nil
+	}
+
+	// if the client didn't set SNI, then we need to inspect the requested IP so that we can choose
+	// a certificate from our list if we specifically handle that IP.  This can happen when a cert
+	// has a common name that is an IP or when it is specifically mapped by name.
```

Comment:
given the SNI RFC explicitly forbids passing IP addresses to select certs, I'm unsure of the implications of this change

https://tools.ietf.org/html/rfc6066#section-3:

> Literal IPv4 and IPv6 addresses are not permitted in "HostName"

The worst-case scenario I can think of would be:
1. the default serving certificate is one understood/trusted by clients (e.g. signed for DNS name "myapiserver.com")
2. the requested IP address does *not* correspond to the address the client requested (e.g. a network path via a load balancer that makes the requested IP appear to be 127.0.0.1)
3. an additional certificate has a common name (or IP SAN) that matches the requested IP address as seen by the API server

In that scenario, this PR would change the API server's behavior from using the default certificate (which would currently work) to using another configured certificate (which would not work).

My questions for someone who knows (sig-network?) would be:
* can we rely on SNI being passed in the client hello for clients using DNS to speak to the API server?
* can we rely on the requested IP as seen by the apiserver for clients using IP addresses to speak to the API server (I would expect not)

Reply:
@liggitt updated logic to prevent the common name from being used if it is an IP: https://github.com/kubernetes/kubernetes/pull/85308/files#diff-515eb628bf780d6460d8bcc2765dd256R83